src/clp_ffi_py/utils.cpp

```diff
@@ -59,4 +62,22 @@ auto get_py_bool(bool is_true) -> PyObject* {
     }
     Py_RETURN_FALSE;
 }
+
+auto handle_traceable_exception(clp::TraceableException& exception) noexcept -> void {
+    if (auto* py_ffi_exception{dynamic_cast<ExceptionFFI*>(&exception)}) {
+        auto& exception_context{py_ffi_exception->get_py_exception_context()};
+        if (exception_context.has_exception() && exception_context.restore()) {
```

Comment:
why we are &&ing `exception_context.has_exception()` and `exception_context.restore()`? Is it possible that the 2nd expression evaluates to false while the first is true?

it looks like exception_context.restore() will not have any effect if it doesn't have an exception, so wonder do we really need `has_exception()` method?

Reply:
1. In the current implementation, if `has_exception` returns false, `restore` will always return true. However, I think it might be better to keep track of `restore`'s return value to ensure that a Python exception has been set before returning.
2. If `has_exception` returns false, `restore` will clear any exceptions already set

Reply:
>If has_exception returns false, restore will clear any exceptions already set

Is this behavior desired in some cases? At least in this line, I can see that you are preventing this from happening? (i.e., you are calling has_exception() first and will not proceed to call restore() if has_exception is false, so that the exception that is already set won;t be cleared

Reply:
1. The point of adding this check is we don't want to restore an empty exception
2. Yes, it will be used if we want to store exception states and recover it later

Reply:
what if we write the restore in this way:

```
[[no_discard]] auto restore() noexcept -> bool {
    auto const exception_has_been_set{has_exception()};
    if (exception_has_been_set) {
      PyErr_Restore(m_type, m_value, m_traceback);
      m_type = nullptr;
      m_value = nullptr;
      m_traceback = nullptr;
    }
    return exception_has_been_set;
}
```

Reply:
This will prevent us from restoring the empty exception.
Check this example: https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_finalize
We may want to recover the exception state before returning, meaning that we may need to restore an empty exception

Reply:
Discussed offline and we decide to move `exception_context.restore()` into the if statement